Symbolic shapes mega merge PR (Oct 3)

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86160

- TensorGeometry supports symint
- check_size supports symint
- functorch batch rule improved symint
- Some operator support for symint in LTC
- More supported operations on SymInt and SymFloat
- More symint support in backwards formulas

This merge includes code contributions from bdhirsh and anjali411.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Symbolic shapes mega merge PR (Oct 3)

- TensorGeometry supports symint
- check_size supports symint
- functorch batch rule improved symint
- Some operator support for symint in LTC
- More supported operations on SymInt and SymFloat
- More symint support in backwards formulas
- Some random sympy changes I got tired of staring at

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Symbolic shapes mega merge PR (Oct 3)"


- TensorGeometry supports symint
- check_size supports symint
- functorch batch rule improved symint
- Some operator support for symint in LTC
- More supported operations on SymInt and SymFloat
- More symint support in backwards formulas
- Some random sympy changes I got tired of staring at

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

Chillee commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -446,8 +499,8 @@ def _maybe_guard_eq(self, expr: "sympy.Eq") -> None:
                 try:
-                    solutions = sympy.solve(lhs - rhs, mod_expr, dict=True)
-                    if len(solutions) == 1 and solutions[0][mod_expr] == 0:
+                    solutions = sympy.solve(lhs - rhs, mod_expr)
```
Don't land this, I think.

ezyang replied:
I'm going to revert this from branch too then

Chillee commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -473,5 +526,6 @@ def evaluate_expr(self, expr: "sympy.Expr"):
         # NB: drop two frames; evaluate_expr and the Sym* function that
         # actually called us
         stack = ''.join(traceback.format_list(traceback.extract_stack()[:-2]))
+        stack += ''.join(torch.utils.get_cpp_backtrace())
```
I think this also slows down tracing quite a bit.

ezyang committed:
Update on "Symbolic shapes mega merge PR (Oct 3)"


- TensorGeometry supports symint
- check_size supports symint
- functorch batch rule improved symint
- Some operator support for symint in LTC
- More supported operations on SymInt and SymFloat
- More symint support in backwards formulas
- Some random sympy changes I got tired of staring at

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Symbolic shapes mega merge PR (Oct 3)"


- TensorGeometry supports symint
- check_size supports symint
- functorch batch rule improved symint
- Some operator support for symint in LTC
- More supported operations on SymInt and SymFloat
- More symint support in backwards formulas
- Some random sympy changes I got tired of staring at

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Symbolic shapes mega merge PR (Oct 3)"


- TensorGeometry supports symint
- check_size supports symint
- functorch batch rule improved symint
- Some operator support for symint in LTC
- More supported operations on SymInt and SymFloat
- More symint support in backwards formulas
- Some random sympy changes I got tired of staring at

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

Chillee commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -177,6 +178,26 @@ def eval(cls, base, divisor):
                     sympy.simplify(base / gcd), sympy.simplify(divisor / gcd)
                 )
 
+    class Ceil(sympy.Function):
```
Why can't we use this here? https://docs.sympy.org/latest/modules/functions/elementary.html

ezyang replied:
^- @bdhirsh

bdhirsh replied:
Oh yeah we should use ‘ceiling’, I must have missed it. I mostly just copied off of how your did floor div

ezyang committed:
Update on "Symbolic shapes mega merge PR (Oct 3)"


- TensorGeometry supports symint
- check_size supports symint
- functorch batch rule improved symint
- Some operator support for symint in LTC
- More supported operations on SymInt and SymFloat
- More symint support in backwards formulas

This merge includes code contributions from bdhirsh and anjali411.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Symbolic shapes mega merge PR (Oct 3)"


- TensorGeometry supports symint
- check_size supports symint
- functorch batch rule improved symint
- Some operator support for symint in LTC
- More supported operations on SymInt and SymFloat
- More symint support in backwards formulas

This merge includes code contributions from bdhirsh and anjali411.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

Chillee reviewed (approved):
symbolic shapes change is unclear to me, but will approve to unblock.

ezyang commented:
@pytorchbot merge
